File: libs/power-log-parser/src/lib/parsers/minion-summoned-parser.ts
import { BlockType, GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Action, FullEntity, Node, ShowEntity, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class MinionSummonedParser implements ActionParser {
	readonly ParserName = 'MinionSummonedParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		const isTriggerPhase =
			node.Parent == null ||
			node.Parent.Type !== Action ||
			(node.Parent.Object as Action).Type === (BlockType.TRIGGER as number);
		const isPowerPhase =
			node.Parent == null ||
			node.Parent.Type !== Action ||
			(node.Parent.Object as Action).Type === (BlockType.POWER as number);
		if (!isTriggerPhase && !isPowerPhase) {
			return false;
		}
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === TagChange &&
			(node.Object as TagChange).Name === (GameTag.ZONE as number) &&
			(node.Object as TagChange).Value === (Zone.PLAY as number) &&
			this.GameState.CurrentEntities.has((node.Object as TagChange).Entity) &&
			this.GameState.CurrentEntities.get((node.Object as TagChange).Entity)!.IsMinionLike()
		);
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		const createFromFullEntity =
			node.Type === FullEntity &&
			(node.Object as FullEntity).GetTag(GameTag.ZONE) === (Zone.PLAY as number) &&
			(node.Object as FullEntity).IsMinionLike();
		const createFromShowEntity =
			node.Type === ShowEntity &&
			(node.Object as ShowEntity).GetTag(GameTag.ZONE) === (Zone.PLAY as number) &&
			(node.Object as ShowEntity).IsMinionLike() &&
			this.GameState.CurrentEntities.has((node.Object as ShowEntity).Entity) &&
			this.GameState.CurrentEntities.get((node.Object as ShowEntity).Entity)!.GetTag(GameTag.ZONE) !==
				(Zone.PLAY as number);
		return stateType === StateType.PowerTaskList && (createFromFullEntity || createFromShowEntity);
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity)!;
		const cardId = entity.CardId;
		const controllerId = entity.GetEffectiveController();
		const creatorEntityId = entity.GetTag(GameTag.CREATOR);
		const creatorEntityCardId = this.GameState.CurrentEntities.has(creatorEntityId)
			? this.GameState.CurrentEntities.get(creatorEntityId)!.CardId
			: null;
		const eventName =
			entity.GetTag(GameTag.ZONE) === (Zone.HAND as number) ? 'MINION_SUMMONED_FROM_HAND' : 'MINION_SUMMONED';
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				eventName,
				GameEventHelper.CreateProvider(eventName, cardId, controllerId, entity.Id, this.StateFacade, {
					CreatorCardId: creatorEntityCardId,
					CreatorEntityId: creatorEntityId,
					Tags: entity.Tags,
				}),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const isPlayBlock =
			node.Parent != null &&
			node.Parent.Type === Action &&
			(node.Parent.Object as Action).Type === (BlockType.PLAY as number);
		if (isPlayBlock) {
			const parentAction = node.Parent!.Object as Action;
			const createdEntityId =
				node.Type === FullEntity
					? (node.Object as FullEntity).Entity
					: (node.Object as ShowEntity).Entity;
			if (createdEntityId === parentAction.Entity) {
				return null;
			}
		}

		if (node.Type === FullEntity) {
			return this.CreateFromFullEntity(node);
		} else {
			return this.CreateFromShowEntity(node);
		}
	}

	private CreateFromFullEntity(node: Node): GameEventProvider[] | null {
		const fullEntity = node.Object as FullEntity;
		const cardId = fullEntity.CardId;
		const controllerId = fullEntity.GetEffectiveController();
		const creatorEntityId = fullEntity.GetTag(GameTag.CREATOR);
		const creatorEntityCardId = this.GameState.CurrentEntities.has(creatorEntityId)
			? this.GameState.CurrentEntities.get(creatorEntityId)!.CardId
			: null;
		return [
			GameEventProvider.Create(
				fullEntity.TimeStamp,
				'MINION_SUMMONED',
				GameEventHelper.CreateProvider(
					'MINION_SUMMONED',
					cardId,
					controllerId,
					fullEntity.Id,
					this.StateFacade,
					{
						CreatorCardId: creatorEntityCardId,
						CreatorEntityId: creatorEntityId,
						Tags: fullEntity.Tags,
					},
				),
				true,
				node,
			),
		];
	}

	private CreateFromShowEntity(node: Node): GameEventProvider[] | null {
		const showEntity = node.Object as ShowEntity;
		const cardId = showEntity.CardId;
		const controllerId = showEntity.GetEffectiveController();
		const creatorEntityId = showEntity.GetTag(GameTag.CREATOR);
		const creatorEntityCardId = this.GameState.CurrentEntities.has(creatorEntityId)
			? this.GameState.CurrentEntities.get(creatorEntityId)!.CardId
			: null;
		const previousZone = this.GameState.CurrentEntities.get(showEntity.Entity)!.GetTag(GameTag.ZONE);
		const eventName = previousZone === (Zone.HAND as number) ? 'MINION_SUMMONED_FROM_HAND' : 'MINION_SUMMONED';
		return [
			GameEventProvider.Create(
				showEntity.TimeStamp,
				eventName,
				GameEventHelper.CreateProvider(
					eventName,
					cardId,
					controllerId,
					showEntity.Entity,
					this.StateFacade,
					{
						CreatorCardId: creatorEntityCardId,
						CreatorEntityId: creatorEntityId,
						Tags: showEntity.Tags,
					},
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/MinionSummonedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class MinionSummonedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public MinionSummonedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            var isTriggerPhase = (node.Parent == null
                      || node.Parent.Type != typeof(Parser.ReplayData.GameActions.Action)
                      || (node.Parent.Object as Parser.ReplayData.GameActions.Action).Type == (int)BlockType.TRIGGER);
            var isPowerPhase = (node.Parent == null
                       || node.Parent.Type != typeof(Parser.ReplayData.GameActions.Action)
                       || (node.Parent.Object as Parser.ReplayData.GameActions.Action).Type == (int)BlockType.POWER);
            if (!isTriggerPhase && !isPowerPhase)
            {
                return false;
            }
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.ZONE
                && (node.Object as TagChange).Value == (int)Zone.PLAY
                && GameState.CurrentEntities.ContainsKey((node.Object as TagChange).Entity)
                && GameState.CurrentEntities[(node.Object as TagChange).Entity].IsMinionLike();
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            var createFromFullEntity = node.Type == typeof(FullEntity)
                && (node.Object as FullEntity).GetTag(GameTag.ZONE) == (int)Zone.PLAY
                && (node.Object as FullEntity).IsMinionLike();
            var createFromShowEntity = node.Type == typeof(ShowEntity)
                && (node.Object as ShowEntity).GetTag(GameTag.ZONE) == (int)Zone.PLAY
                && (node.Object as ShowEntity).IsMinionLike()
                && GameState.CurrentEntities.ContainsKey((node.Object as ShowEntity).Entity)
                && GameState.CurrentEntities[(node.Object as ShowEntity).Entity].GetTag(GameTag.ZONE) != (int)Zone.PLAY;
            var timeWarped = node.Type == typeof(FullEntity) 
                && GameState.GetGameEntity()?.GetTag(GameTag.BACON_ALT_TAVERN_IN_PROGRESS) == 1
                && (node.Object as FullEntity).GetTag(GameTag.ZONE) == (int)Zone.SETASIDE
                && (node.Object as FullEntity).IsMinionLike()
                && (node.Object as FullEntity).GetTag(GameTag.BACON_TIMEWARPED) == 1;
            return stateType == StateType.PowerTaskList
                && (createFromFullEntity || createFromShowEntity);
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities[(node.Object as TagChange).Entity];
            var cardId = entity.CardId;
            var controllerId = entity.GetEffectiveController();
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, null, null);
            var creatorEntityId = entity.GetTag(GameTag.CREATOR);
            var creatorEntityCardId = GameState.CurrentEntities.ContainsKey(creatorEntityId)
                ? GameState.CurrentEntities[creatorEntityId].CardId
                : null;
            var eventName = entity.GetTag(GameTag.ZONE) == (int)Zone.HAND ? "MINION_SUMMONED_FROM_HAND" : "MINION_SUMMONED";
            //var shouldShortCircuit = ShouldShortCircuit(node);
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                eventName,
                GameEvent.CreateProvider(
                    eventName,
                    cardId,
                    controllerId,
                    entity.Id,
                    StateFacade,
                    //gameState,
                    new {
                        CreatorCardId = creatorEntityCardId,
                        CreatorEntityId = creatorEntityId,
                        Tags = entity.Tags,
                    }
                ),
                true,
                node
            )};
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            // Prevent the same minion from being created twice, once from the PLAY block
            // and once from the Entity block
            var isPlayBlock = (node.Parent != null
                    && node.Parent.Type == typeof(Parser.ReplayData.GameActions.Action)
                    && (node.Parent.Object as Parser.ReplayData.GameActions.Action).Type == (int)BlockType.PLAY);
            if (isPlayBlock)
            {
                var parentAction = node.Parent.Object as Parser.ReplayData.GameActions.Action;
                var createdEntityId = node.Type == typeof(FullEntity)
                    ? (node.Object as FullEntity).Entity
                    : (node.Object as ShowEntity).Entity;
                if (createdEntityId == parentAction.Entity)
                {
                    return null;
                }
            }

            if (node.Type == typeof(FullEntity))
            {
                return CreateFromFullEntity(node);
            }
            else
            {
                return CreateFromShowEntity(node);
            }
        }

        public List<GameEventProvider> CreateFromFullEntity(Node node)
        {
            var fullEntity = node.Object as FullEntity;
            var cardId = fullEntity.CardId;
            var controllerId = fullEntity.GetEffectiveController();
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, null, null);
            var creatorEntityId = fullEntity.GetTag(GameTag.CREATOR);
            var creatorEntityCardId = GameState.CurrentEntities.ContainsKey(creatorEntityId)
                ? GameState.CurrentEntities[creatorEntityId].CardId
                : null;
            //var shouldShortCircuit = ShouldShortCircuit(node);
            return new List<GameEventProvider>
            {
                GameEventProvider.Create(
                fullEntity.TimeStamp,
                "MINION_SUMMONED",
                GameEvent.CreateProvider(
                    "MINION_SUMMONED",
                    cardId,
                    controllerId,
                    fullEntity.Id,
                    StateFacade,
                    //gameState,
                    new
                    {
                        CreatorCardId = creatorEntityCardId,
                        CreatorEntityId = creatorEntityId,
                        Tags = fullEntity.Tags,
                    }
                ),
                true,
                node
            )};
        }

        public List<GameEventProvider> CreateFromShowEntity(Node node)
        {
            var showEntity = node.Object as ShowEntity;
            var cardId = showEntity.CardId;
            var controllerId = showEntity.GetEffectiveController();
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, null, null);
            var creatorEntityId = showEntity.GetTag(GameTag.CREATOR);
            var creatorEntityCardId = GameState.CurrentEntities.ContainsKey(creatorEntityId)
                ? GameState.CurrentEntities[creatorEntityId].CardId
                : null;
            var previousZone = GameState.CurrentEntities[(node.Object as ShowEntity).Entity].GetTag(GameTag.ZONE);
            var eventName = previousZone == (int)Zone.HAND ? "MINION_SUMMONED_FROM_HAND" : "MINION_SUMMONED";
            //var shouldShortCircuit = ShouldShortCircuit(node);
            return new List<GameEventProvider> { GameEventProvider.Create(
                showEntity.TimeStamp,
                eventName,
                GameEvent.CreateProvider(
                    eventName,
                    cardId,
                    controllerId,
                    showEntity.Entity,
                    StateFacade,
                    //gameState,
                    new {
                        CreatorCardId = creatorEntityCardId,
                        CreatorEntityId = creatorEntityId,
                        Tags = showEntity.Tags,
                    }
                ),
                true,
                node
            )};
        }

        // Only short-circuit after a reroll
        //private bool ShouldShortCircuit(Node node)
        //{
        //    if (!StateFacade.IsBattlegrounds())
        //    {
        //        return false;
        //    }

        //    if (node.Parent == null || node.Parent.Type != typeof(Action))
        //    {
        //        return false;
        //    }

        //    var action = node.Parent.Object as Action;
        //    return action.Type == (int)BlockType.POWER
        //        && GameState.CurrentEntities.ContainsKey(action.Entity)
        //        && (GameState.CurrentEntities[action.Entity].CardId == CardIds.Refresh_TB_BaconShop_8p_Reroll_Button
        //            || GameState.CurrentEntities[action.Entity].CardId == CardIds.Refresh_TB_BaconShop_1p_Reroll_Button);
        //}
    }
}
